File: quick_sort.py
import csv
import time

class CsvReader:
    class DataPair:
        def __init__(self, id_: int, value: str):
            self.id = int(id_)
            self.value = value

        def __repr__(self):
            return f"({self.id},{self.value})"

    def read_csv(self, filepath):
        data = []
        with open(filepath, newline='', encoding='utf-8') as csvfile:
            reader = csv.reader(csvfile)
            for row in reader:
                if len(row) >= 2:
                    data.append(self.DataPair(row[0], row[1]))
        return data

    def write_values_to_txt(self, data, filename):
        with open(filename, 'w', encoding='utf-8') as f:
            for pair in data:
                f.write(f"{pair.id},{pair.value}\n")


class QuickSort:
    def quick_sort(self, data):
        if len(data) <= 1:
            return data
        pivot_index = len(data) - 1
        pivot = data[pivot_index]
        left = []
        right = []

        for i in range(pivot_index):
            if data[i].id <= pivot.id:
                left.append(data[i])
            else:
                right.append(data[i])

        sorted_data = []
        sorted_data.extend(self.quick_sort(left))
        sorted_data.append(pivot)
        sorted_data.extend(self.quick_sort(right))
        return sorted_data


def main():
    reader = CsvReader()
    try:
        csv_file = input("Enter the path to the CSV file: ")
        data = reader.read_csv(csv_file)
    except Exception as e:
        print(f"Error reading file: {e}")
        return

    if data:
        sorter = QuickSort()
        start_time = time.time_ns()
        data = sorter.quick_sort(data)
        end_time = time.time_ns()
        print(f"Execution time: {end_time - start_time} ns")

        try:
            reader.write_values_to_txt(data, "output_quicksort.txt")
        except Exception as e:
            print(f"Error writing to file: {e}")
    else:
        print("No data to display.")


if __name__ == "__main__":
    main()

File: quick_sort.java
import utils.CsvReader;
import utils.CsvReader.DataPair;
import java.util.List;
import java.util.ArrayList;
import java.io.IOException;
import java.util.Scanner;

public class quick_sort {
	public List<DataPair> QuickSort(List<DataPair> data){
		
		if (data.size() <= 1){
			return data;
		}
		int pivotIndex = data.size() - 1;
		DataPair pivot = data.get(pivotIndex);
		List<DataPair> left = new ArrayList<>();
		List<DataPair> right = new ArrayList<>();
		for (int i = 0; i < pivotIndex; i++){
			if (data.get(i).id <= pivot.id){
				left.add(data.get(i));
			} else {
				right.add(data.get(i));
			}
		}
		List<DataPair> sorted = new ArrayList<>();
		sorted.addAll(QuickSort(left));
		sorted.add(pivot);
		sorted.addAll(QuickSort(right));
		return sorted;
	}

	public static void main(String[] args) {
		CsvReader reader = new CsvReader();
		
		List<DataPair> data = null;
		try (Scanner scanner = new Scanner(System.in)) {
			System.out.print("Enter the path to the CSV file: ");
			String csvFile = scanner.nextLine();
			data = reader.readCsv(csvFile);
			// Now you can merge-sort this data
		} catch (IOException e) {
			System.err.println("Error reading file: " + e.getMessage());
		}
		
		if (data != null) {

			// System.out.println("ID" + data.get(2).id + ", Value: " + data.get(2).value);
			long startTime = System.nanoTime();
			data = new quick_sort().QuickSort(data);
			long endTime = System.nanoTime();
			float timeTaken = (float) (endTime - startTime)/1000000000;
			System.out.println("Execution time: " + timeTaken + " seconds");
			
			try {
				reader.writeValuesToTxt(data, "output_quicksort.txt");
			} catch (IOException e) {
				System.err.println("Error writing to file: " + e.getMessage());
			}
		} else {
			System.out.println("No data to display.");
		}
		}
}
